aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,953 @@
+        # Add 1 so line numbering starts at 1
+        mapping['__line__'] = node.start_mark.line + 1
+        return mapping
+
+# Parse native_functions.yaml into a sequence of NativeFunctions
+def parse_native_yaml(path: str) -> List[NativeFunction]:
+    with open(path, 'r') as f:
+        es = yaml.load(f, Loader=LineLoader)
+    assert isinstance(es, list)
+    rs: List[NativeFunction] = []
+    for e in es:
+        assert isinstance(e.get('__line__'), int), e
+        loc = Location(path, e['__line__'])
+        funcs = e.get('func')
+        with context(f'in {loc}:\n  {funcs}'):
+            rs.append(NativeFunction.from_yaml(e, loc))
+    return rs
+
+native_functions = parse_native_yaml('aten/src/ATen/native/native_functions.yaml')
+# pprint.pprint([dataclasses.asdict(f) for f in native_functions])
+
+# TODO: TensorOptions argument detection
+# TODO: Extra enforcement of inplace functions having mutable self
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                           CODE GENERATION
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# This is not fully constructed yet, but eventually this will actually
+# do the code generation
+
+TEMPLATE_PATH = "aten/src/ATen/templates"
+
+# When we interpret things as C++ types, there are a bunch of
+# different modalities we have to consider
+#
+# - Return versus argument type
+# - Mutable type (inplace, out argument)
+# - Public API versus internal calling convention versus legacy calling
+#   convention
+#
+# I'm not really sure how to structure this logic yet, but here is a
+# sketch.  This function is ONLY correct for CPUType.h at the moment;
+# I bet I am going to need another parameter before I'm done
+def cpp_type(t: Type, *, mutable: bool, argument: bool, legacy_optional: bool) -> str:
+    if isinstance(t, BaseType):
+        if t.name == BaseTy.Tensor:
+            if mutable:
+                return 'Tensor &'
+            else:
+                if argument:
+                    return 'const Tensor &'
+                else:
+                    return 'Tensor'
+        elif t.name == BaseTy.int:
+            return 'int64_t'
+        elif t.name == BaseTy.float:
+            return 'double'
+        elif t.name == BaseTy.str:
+            return 'std::string'
+        elif t.name in [BaseTy.bool, BaseTy.QScheme, BaseTy.Scalar,
+                BaseTy.ScalarType, BaseTy.Generator, BaseTy.Storage,
+                BaseTy.Layout, BaseTy.Device, BaseTy.MemoryFormat,
+                BaseTy.Dimname, BaseTy.ConstQuantizerPtr]:
+            # These C++ names coincidentally line up with their schema
+            # names
+            return t.name.name
+        else:
+            assert False, f"unsupported type: {t}"
+    elif isinstance(t, OptionalType):
+        # TODO: these arguments are smoothed over by the hacky wrapper
+        if argument and legacy_optional and str(t.elem) == 'Tensor':
```

Comment:
bhosmer suggests factoring out the `legacy_optional` logic, perhaps into a wrapper